compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -193,7 +193,7 @@ fn layout_of_uncached<'tcx>(
         }
 
         ty::Dynamic(_, _, ty::DynStar) => {
-            let mut data = scalar_unit(Int(dl.ptr_sized_integer(), false));
+            let mut data = scalar_unit(Pointer(AddressSpace::DATA));
```

Comment:
This seems to cause issues.
```
---- [ui] tests/ui/dyn-star/box.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 101
command: "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "/home/r/src/rust/rustc/tests/ui/dyn-star/box.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/home/r/src/rust/rustc/tests/ui=fake-test-src-base" "-C" "prefer-dynamic" "-o" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/test/ui/dyn-star/box/a" "-Crpath" "-Cdebuginfo=0" "-Lnative=/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/test/ui/dyn-star/box/auxiliary" "-C" "opt-level=0"
stdout: none
--- stderr -------------------------------
Invalid InsertValueInst operands!
  %15 = insertvalue { ptr, ptr } undef, i64 %14, 0
in function _ZN3box13make_dyn_star17h0c15bb343f1925bdE
LLVM ERROR: Broken function found, compilation aborted!
```
Probably somewhere in the backend something still assumes an integer type here, and inserts an `i64` into this `{ ptr, ptr }`? This actually casts a Box to `dyn*`, so the input value *is* a pointer -- no idea where the `i64` comes from.

Reply:
Looking at the LLVM IR on current nightly
```
  %10 = ptrtoint ptr %_4.i to i64, !dbg !1309
  %11 = insertvalue { i64, ptr } undef, i64 %10, 0, !dbg !1310
  %12 = insertvalue { i64, ptr } %11, ptr @vtable.1, 1, !dbg !1310
```
We are generating a `ptrtoint`?!? That doesn't sound right at all. We shouldn't be generating such a cursed operation unless the user actually asked for it.

Reply:
Oh, I just found https://github.com/rust-lang/rust/pull/102641...

We should be doing the opposite: if the input is an integer, we should cast/transmute/bit/cast/whatever it to a pointer.

Reply:
Strangely just removing that pointer type hack seems to work? I would have expected errors from where we cast usize to dyn* (I think we are now inserting an `i64` where a `ptr` should go) but strangely, there are no such errors...

Reply:
Ah, on CI it does fail. CI uses an older, pre-`ptr` version of LLVM.